File: update_sitemap.py
import json

ARTICLES_PATH = r"c:\My projects\PettyCash_Static_Web\js\articles-data.js"
SITEMAP_PATH = r"c:\My projects\PettyCash_Static_Web\sitemap.xml"

with open(ARTICLES_PATH, "r", encoding="utf-8") as f:
    articles_text = f.read()

# Extract JSON array from JS file
start = articles_text.find("[")
end = articles_text.rfind("]")
articles_data = json.loads(articles_text[start:end+1])

articles = [(a["slug"], a["date"]) for a in articles_data]
print(f"Found {len(articles)} articles in data file")

with open(SITEMAP_PATH, "r", encoding="utf-8") as f:
    sitemap = f.read()

# Find the position of the first article URL
first_article_idx = sitemap.find("https://www.pettycash.site/article.html")
if first_article_idx == -1:
    # Insert before closing urlset
    insert_idx = sitemap.rfind("</urlset>")
else:
    # Find the start of that <url> block
    insert_idx = sitemap.rfind("<url>", 0, first_article_idx)

before = sitemap[:insert_idx]
after = sitemap[sitemap.rfind("</urlset>"):]

article_urls = []
for slug, date in articles:
    article_urls.append(f"""  <url>
    <loc>https://www.pettycash.site/article.html?slug={slug}</loc>
    <lastmod>{date}</lastmod>
    <changefreq>monthly</changefreq>
    <priority>0.6</priority>
  </url>""")

new_sitemap = before + "\n".join(article_urls) + "\n" + after

with open(SITEMAP_PATH, "w", encoding="utf-8") as f:
    f.write(new_sitemap)

print(f"Updated sitemap with {len(articles)} article URLs")

File: update-sitemap.js
const fs = require('fs');
const path = require('path');

const root = __dirname;
const baseUrl = 'https://www.pettycash.site';
const today = new Date().toISOString().split('T')[0];

// Static pages to always include
const staticPages = [
  { loc: `${baseUrl}/tools/`, lastmod: today, changefreq: 'weekly', priority: '0.8' },
  { loc: `${baseUrl}/tools/saudi-invoice-generator.html`, lastmod: today, changefreq: 'monthly', priority: '0.7' },
  { loc: `${baseUrl}/tools/receipt-generator.html`, lastmod: today, changefreq: 'monthly', priority: '0.7' },
  { loc: `${baseUrl}/tools/petty-cash-voucher-generator.html`, lastmod: today, changefreq: 'monthly', priority: '0.7' },
  { loc: `${baseUrl}/`, lastmod: today, changefreq: 'weekly', priority: '1.0' },
  { loc: `${baseUrl}/features.html`, lastmod: today, changefreq: 'monthly', priority: '0.8' },
  { loc: `${baseUrl}/pricing.html`, lastmod: today, changefreq: 'monthly', priority: '0.8' },
  { loc: `${baseUrl}/blog.html`, lastmod: today, changefreq: 'weekly', priority: '0.7' },
  { loc: `${baseUrl}/help.html`, lastmod: today, changefreq: 'monthly', priority: '0.7' },
  { loc: `${baseUrl}/about.html`, lastmod: today, changefreq: 'monthly', priority: '0.7' },
  { loc: `${baseUrl}/contact.html`, lastmod: today, changefreq: 'monthly', priority: '0.7' },
];

// Extract articles from articles-data.js
const articlesFile = fs.readFileSync(path.join(root, 'js/articles-data.js'), 'utf8');
const articles = [];
let currentArticle = null;

for (const line of articlesFile.split('\n')) {
  const slugMatch = line.match(/"slug"\s*:\s*"([^"]+)"/);
  if (slugMatch) {
    if (currentArticle) articles.push(currentArticle);
    currentArticle = { slug: slugMatch[1], date: today };
  }
  const dateMatch = line.match(/"date"\s*:\s*"([^"]+)"/);
  if (dateMatch && currentArticle) {
    currentArticle.date = dateMatch[1];
  }
}
if (currentArticle) articles.push(currentArticle);

// Remove duplicates by slug
const seen = new Set();
const uniqueArticles = [];
for (const a of articles) {
  if (!seen.has(a.slug)) {
    seen.add(a.slug);
    uniqueArticles.push(a);
  }
}

const articleUrls = uniqueArticles.map(a => ({
  loc: `${baseUrl}/article.html?slug=${a.slug}`,
  lastmod: a.date,
  changefreq: 'monthly',
  priority: '0.6'
}));

const urls = [...staticPages, ...articleUrls];

const xml = `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
${urls.map(u => `  <url>
    <loc>${u.loc}</loc>
    <lastmod>${u.lastmod}</lastmod>
    <changefreq>${u.changefreq}</changefreq>
    <priority>${u.priority}</priority>
  </url>`).join('\n')}
</urlset>
`;

fs.writeFileSync(path.join(root, 'sitemap.xml'), xml);
console.log(`Updated sitemap.xml with ${staticPages.length} static pages and ${articleUrls.length} articles.`);
console.log(`Total URLs: ${urls.length}`);
